README.md
# Telegram Bot Compiler

## Description

Own development of a program for building your own telegram bot in Python. At the moment, you can add buttons through the program menu and set up actions for them, everything is easy and simple, but this is early development and there will be new improvements and features in the future.

## Features

- Create and customize buttons with different names and types.
- Assign actions to buttons, such as opening programs or executing commands.
- Organize buttons into columns for better organization and management.
- Toggle visibility of buttons with a single click.
- Dark mode option for a visually pleasing experience.
- Multi-language support for internationalization.

## Contributing

Contributions are welcome! If you would like to contribute to the Telegram Bot Compiler, please follow these steps:

1. Fork the repository.
2. Create a new branch.
3. Make your changes and commit them.
4. Push your changes to your forked repository.
5. Submit a pull request.

## License

The Telegram Bot Compiler is released under the [MIT License](https://opensource.org/licenses/MIT).

## Acknowledgements

We would like to acknowledge the following resources for their contributions to this project:

- Bootstrap CSS framework
- Font Awesome icon library
- jQuery JavaScript library
- Eel Python library

## Contact

For any inquiries or support regarding the TBC, please contact:

Email: [levroemail@gmail.com](mailto:levroemail@gmail.com)
Telegram: [t.me/xwkwx](https://t.me/xwkwx)

## Screenshots

![Screenshot 1](https://github.com/LevrikM/LevrikM.github.io/assets/48616642/07ef6fd2-b82d-4011-885b-cc3e01ae410d)
![Screenshot 2](https://github.com/LevrikM/telegramBotCompiler/assets/48616642/9ad3f774-cfca-4719-94d5-7c76c0ce0ef6)
![Screenshot 3](https://github.com/LevrikM/telegramBotCompiler/assets/48616642/4145c315-b1f0-4288-9290-f67cea45cb47)


CHANGELOG.md
# Changelog

All notable changes to the Telegram Bot Compiler project will be documented in this file.

## [1.1.0] - 2023-06-09

### Added
- Save button list display state.
- Hide the full list of buttons if there are none and vice versa.

### Changed
- Conducted initial code refactoring for a better overview of the code
- Changed minor parts of the interface and icons for better performance.

### Fixed
- Fixed bot and logo settings window icon
- Resolved minor bugs and issues.

### Removed
- Removed the 'folder' field in the button file for the bot due to its non-use in the program.


## [1.1.1] - 2023-06-09

### Added
- Hide the full list of tokens if there are none and vice versa.
- Added a button to hide the active bot token in settings with state saving

### Changed
- The hide buttons button has been changed to a more convenient and understandable one

### Fixed
- Fixed display of Cyrillic in Telegram bot and saving to file


Old:  
<img src="https://github.com/LevrikM/telegramBotCompiler/assets/48616642/359ac904-8538-4655-9e18-17f0a1eac5ad" alt="image 1" style="width:700px;">
<br>
New:  
<img src="https://github.com/LevrikM/telegramBotCompiler/assets/48616642/1ffdc7af-b672-401a-87be-4298a24a36a7" alt="image 2" style="width:700px;">


main.py
import json
import eel

eel.init('web/main', allowed_extensions=['.js', '.html', '.css'])

@eel.expose
def save_token(token):
    saved_tokens = get_saved_tokens()
    if token not in saved_tokens:
        saved_tokens.append(token)
        save_tokens(saved_tokens)
        print(f'Token: {token} saved!')
        return "success"
    else:
        print(f'Token: {token} already exists!')
        return "exists"

def get_saved_tokens():
    try:
        with open('tokens.json', 'r') as file:
            saved_tokens = json.load(file)
    except (FileNotFoundError, json.JSONDecodeError):
        saved_tokens = []
    return saved_tokens

def save_tokens(saved_tokens):
    with open('tokens.json', 'w') as file:
        json.dump(saved_tokens, file)

@eel.expose
def delete_token(token):
    saved_tokens = get_saved_tokens()
    if token in saved_tokens:
        saved_tokens.remove(token)
        save_tokens(saved_tokens)
        print(f'Token: {token} deleted!')
    else:
        print(f'Token: {token} does not exist!')

@eel.expose
def get_saved_tokens_js():
    saved_tokens = get_saved_tokens()
    return saved_tokens

@eel.expose
def save_settings(language, theme):
    settings = {'language': language, 'darkMode': theme}

    with open('settings.json', 'w') as file:
        json.dump(settings, file)

    print(f'Settings saved: [language: {language}, darkMode: {theme}]')

@eel.expose
def get_settings():
    try:
        with open('settings.json', 'r') as file:
            settings = json.load(file)
    except FileNotFoundError:
        settings = {'language': 'en', 'darkMode': False}  # default

    return settings

@eel.expose
def print_to_console(text):
    print(f'TokenWindows out: {text}')


import subprocess
import sys
@eel.expose
def run_program(token):
    print(f'Running program on token: {token}')
    subprocess.call(['python', 'bot.py', token])
    eel.expose(sys.exit) 
    sys.exit() 

eel.start('main.html', mode="chrome", size=(1300, 600), resizable=False)


web/main/script.js
function saveToken() {
    var token = document.getElementById("tokenInput").value;
    var errorMessage = document.getElementById("errorMessage");
    if (token.trim() === "") {
        errorMessage.textContent = "Token cannot be empty";
    } else if(token.length < 46){
        errorMessage.textContent = "Token must be 46 characters long"
    }
    else {
        eel.save_token(token)(function (response) {
            if (response === "exists") {
                errorMessage.textContent = "Token already exists";
            } else {
                errorMessage.textContent = "";
                updateTokensList();
            }
        });
    }
}


function deleteToken(token) {
    eel.delete_token(token);
    updateTokensList();
}


function updateTokensList() {
    eel.get_saved_tokens_js()(function(tokens) {
        var tokenList = document.getElementById("tokenList");
        var savedTokens = document.getElementById("savedTokens");
        if(tokens.length == 0){
            savedTokens.style.display = "none";
        }
        else{
            savedTokens.style.display = "block";
            tokenList.innerHTML = "";
            for (var i = 0; i < tokens.length; i++) {
                var token = tokens[i];
                var listItem = document.createElement("li");
                listItem.className = "list-group-item d-flex justify-content-between align-items-center mb-2";
                listItem.innerHTML = `
                    <span>${token}</span>
                    <div>
                        <button class="btn btn-sm btn-primary mr-2" onclick="copyToken('${token}')">
                            <i class="fas fa-copy"></i>
                        </button>
                        <button class="btn btn-sm btn-danger mr-2" onclick="deleteToken('${token}')">
                            <i class="fas fa-trash-alt"></i>
                        </button>
                        <button class="btn btn-sm btn-success" onclick="runProgram('${token}')">
                            <i class="fas fa-play"></i>
                        </button>
                    </div>`;
                tokenList.appendChild(listItem);
            }
        }
        
    });
}


function runProgram(token) {
    eel.run_program(token)();
    window.close()
}

function copyToken(token) {
    var textarea = document.createElement("textarea");
    textarea.value = token;
    document.body.appendChild(textarea);
    textarea.select();
    document.execCommand("copy");
    document.body.removeChild(textarea);
    alert("Token copied to clipboard!");
}

updateTokensList();

////////////////////

var translations = {
    en: {
        titleNav: "Tokens settings",
        title: "Tokens add",
        tokenLabel: "Enter token from BotFather:",
        saveButton: "Save token",
        savedTokens: "Saved Tokens:",
        languageLabel: "Select language:",
        placeholder : "Enter token"
    },
    ua: {
        titleNav : "Налаштування токенів",
        title: "Додавання токенів",
        tokenLabel: "Введіть токен від BotFather:",
        saveButton: "Зберегти токен",
        savedTokens: "Збережені токени:",
        languageLabel: "Оберіть мову:",
        placeholder : "Введіть токен"
    },
    ru: {
        titleNav : "Настройки токенов",
        title: "Добавление токенов",
        tokenLabel: "Введите токен от BotFather:",
        saveButton: "Сохранить токен",
        savedTokens: "Сохраненные токены:",
        languageLabel: "Выберите язык:",
        placeholder : "Введите токен"
    }
};

let selectedLanguage = "";
var darkModeEnabled = "";

var toggleDarkModeButton = $("#toggleDarkModeButton");
var toggleDarkModeIcon = $("#toggleDarkModeIcon");

toggleDarkModeButton.click(function () {
    toggleDarkMode();
});

function updateTexts() {
    var translation = translations[selectedLanguage] || translations.en;

    $("[data-trans]").each(function() {
        var transKey = $(this).data("trans");
        if (translation.hasOwnProperty(transKey)) {
            $(this).text(translation[transKey]);
        }
    });

    $("#tokenInput").attr("placeholder", translation["placeholder"]);
}

function changeLanguage(language){
    selectedLanguage = language;
    eel.save_settings(selectedLanguage, darkModeEnabled)
    updateTexts();
}

function toggleDarkMode() {
    darkModeEnabled = !darkModeEnabled;
    if (darkModeEnabled) {
        $("body").addClass("dark-mode");
        toggleDarkModeIcon.removeClass("fa-adjust").addClass("fa-sun");
        eel.save_settings(selectedLanguage, true)
    } else {
        $("body").removeClass("dark-mode");
        toggleDarkModeIcon.removeClass("fa-sun").addClass("fa-adjust");
        eel.save_settings(selectedLanguage, false)
    }
}


window.onresize = function () {
    window.resizeTo(1300, 600);
}

window.onload = function() {
    eel.get_settings()(function(settings) {
        selectedLanguage = settings.language
        darkModeEnabled = settings.darkMode;
        eel.print_to_console('language: ' + selectedLanguage)
        eel.print_to_console('darkMode: ' + darkModeEnabled)


        if (darkModeEnabled) {
            $("body").addClass("dark-mode");
            toggleDarkModeIcon.removeClass("fa-adjust").addClass("fa-sun");
            
        } else {
            $("body").removeClass("dark-mode");
            toggleDarkModeIcon.removeClass("fa-sun").addClass("fa-adjust");
        }
        // eel.print_to_console('selectedlanguage2: ' + selectedLanguage)
        $("#languageSelect").val(selectedLanguage);
        updateTexts();
    });

}

bot.py
import sys
import eel
import json
import os
import io
import subprocess
import pyautogui
from datetime import datetime
from telegram import ReplyKeyboardMarkup, KeyboardButton, InputFile
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters


eel.init('web/bot', allowed_extensions=['.js', '.html', '.css'])

token = sys.argv[1]
print(f"Received token: {token}")

@eel.expose
def save_settings(language, theme, showbtn, tokenVisible):
    settings = {'language': language, 'darkMode': theme, 'showBtn' : showbtn, 'tokenVisible': tokenVisible}
    with open('settings.json', 'w') as file:
        json.dump(settings, file)

    print(f'Settings saved: [language: {language}, darkMode: {theme}, showBtn: {showbtn}, tokenVisible: {tokenVisible}]')

@eel.expose
def get_settings():
    try:
        with open('settings.json', 'r') as file:
            settings = json.load(file)
    except FileNotFoundError:
        settings = {'language': 'en', 'darkMode': False, 'showBtn': True, 'tokenVisible' : True}  # default

    return settings

@eel.expose
def print_to_console(text):
    print(f'BotWindow out: {text}')

##
@eel.expose
def add_button(button_name, button_type, program_name):
    button = {
        "name": button_name,
        "type": button_type,
        "program": program_name if button_type == "openProgram" else "",
        # "folder": folder_path if button_type == "openFolder" else ""
    }

    buttons_folder = "./web/bot/buttons"
    os.makedirs(buttons_folder, exist_ok=True)
    buttons_file = os.path.join(buttons_folder, f"buttons_{token[:5]}.json")

    if os.path.exists(buttons_file):
        with open(buttons_file, encoding="utf-8") as f:
            buttons_data = json.load(f)
    else:
        buttons_data = []

    if any(button["name"] == button_name for button in buttons_data):
        return "exist"


    buttons_data.append(button)

    with open(buttons_file, "w", encoding="utf-8") as f:
        json.dump(buttons_data, f, ensure_ascii=False)

@eel.expose
def get_buttons():
    buttons_folder = "./web/bot/buttons"
    buttons_file = os.path.join(buttons_folder, f"buttons_{token[:5]}.json")

    if os.path.exists(buttons_file):
        with open(buttons_file, encoding="utf-8") as f:
            buttons_data = json.load(f)
    else:
        buttons_data = []

    return buttons_data

@eel.expose
def delete_button(button_name):
    buttons_folder = "./web/bot/buttons"
    buttons_file = os.path.join(buttons_folder, f"buttons_{token[:5]}.json")

    if os.path.exists(buttons_file):
        with open(buttons_file, encoding="utf-8") as f:
            buttons_data = json.load(f)
        
        filtered_buttons = [button for button in buttons_data if button["name"] != button_name]

        with open(buttons_file, "w", encoding="utf-8") as f:
            json.dump(filtered_buttons, f, ensure_ascii=False)
##
@eel.expose
def return_token():
    # print(f'Returned token: {token}')
    return token

def load_buttons_from_file():
    buttons_file = f'./web/bot/buttons/buttons_{token[:5]}.json'
    buttons = []
    with open(buttons_file, 'r', encoding="utf-8") as file:
        buttons_data = json.load(file)
        if isinstance(buttons_data, list):
            buttons = buttons_data
    return buttons



def get_language():
    settings_file = 'settings.json'
    try:
        with open(settings_file, 'r') as file:
            settings = json.load(file)
            return settings.get('language', 'en')
    except FileNotFoundError:
        return 'en'


def get_message_text(key, language):
    with open(f'./web/bot/messages/messages_{language}.json', 'r',  encoding="utf-8") as file:
        messages = json.load(file)
    return messages.get(key, '')

updater = Updater(token, use_context=True)

def start(update, context):
    language = get_language()
    reply_markup = create_menu_buttons()
    message_text = get_message_text('start_message', language)

    update.message.reply_text(message_text, reply_markup=reply_markup)

######
def create_menu_buttons():
    buttons = load_buttons_from_file()
    language = get_language()
    reload_button = KeyboardButton(get_message_text('reload_buttons_btn', language))
    keyboard = [[reload_button]]
    for button in buttons:
        button_name = button['name']
        keyboard.append([KeyboardButton(button_name)])
    
    reply_markup = ReplyKeyboardMarkup(keyboard, resize_keyboard=True)
    return reply_markup


def open_program(program):
    try:
        subprocess.Popen(program)
        print(f'Open program: {program}')
    except Exception as e:
        print(f'Error open program: {e}')


def reload_buttons():
    return create_menu_buttons()

def button_handler(update, context):
    user_input = update.message.text
    buttons = load_buttons_from_file()
    language = get_language()

    for button in buttons:
        if button['name'] == user_input:
            button_type = button['type']
            if button_type == 'openProgram':
                program_name = button['program']
                open_program(program_name)
                message_text = get_message_text('program_opened_message', language).format(program_name=program_name)
                update.message.reply_text(message_text)
            elif button_type == 'takeScreen':
                screenshot = pyautogui.screenshot()
                screenshot_io = io.BytesIO()
                screenshot.save(screenshot_io, format='PNG')
                screenshot_io.seek(0)
                ##
                now = datetime.now()
                dt_date = now.strftime("%d/%m/%Y")
                dt_time = now.strftime("%H:%M:%S")
                ##
                update.message.reply_text(get_message_text('send_screen', language).format(date=dt_date, time=dt_time))
                update.message.reply_photo(photo=InputFile(screenshot_io, filename='screenshot.png'))
            else:
                message_text = get_message_text('button_selected_message', language).format(button_name=user_input)
                update.message.reply_text(message_text)
            return
    
    if user_input == 'Reload buttons' or user_input == 'Перезавантажити кнопки' or user_input == 'Перезагрузить кнопки':
        reply_markup = reload_buttons()
        message_text = get_message_text('buttons_reloaded_message', language)
        update.message.reply_text(message_text, reply_markup=reply_markup)
    else:
        message_text = get_message_text('unknown_option_message', language)
        update.message.reply_text(message_text)


@eel.expose
def stop_bot():
    try: 
        updater.stop()
        updater.is_idle = False
        return "success"
    except:
        return "error"


@eel.expose
def run_bot():
    try: 
        dispatcher = updater.dispatcher
        start_handler = CommandHandler('start', start)
        dispatcher.add_handler(start_handler)
        dispatcher.add_handler(CommandHandler('reload_buttons', reload_buttons))
        dispatcher.add_handler(MessageHandler(Filters.text, button_handler))
        print(f"Bot start polling at token {token}")
        updater.start_polling()
        return "success"
    except:
        return "error"

eel.start('bot.html', mode="chrome", size=(1300, 600), resizable=False, port=8080)

